feat(builtin): Add `cue fmt` diagnostic

johnallen3d opened the pull request:
Takes advantage of the fact that `cue fmt` returns errors to `stdout` when invalid syntax is encountered and provides them as diagnostics. [CUE](https://cuelang.org/)

I tried using helpers (e.g. `from_pattern`) to reduce lines of code however I couldn't come up with pattern to match the output from `cue fmt`. Each issue spans two lines, first the error message then the position.

```sh
expected label or ':', found 'INT' 42:
    ./foo.cue:3:2
missing ',' in struct literal:
    ./foo.cue:5:7
```

johnallen3d commented:
I could use some help here @jose-elias-alvarez. I have a version of this locally (not builtin, no helpers) that's working however when I convert it to a builtin it is not. I tried adding a spec to identify where the issue is however, as far as I can tell, my `on_output` function is not being called.

jose-elias-alvarez commented on lua/null-ls/builtins/diagnostics/cue_fmt.lua:
```diff
@@ -0,0 +1,46 @@
+        check_exit_code = function(code)
+            return code <= 1
+        end,
+        on_output = function(params)
```
Your issue is most likely that setting `format = "raw"` requires returning diagnostics via the `done` callback, described [here](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/HELPERS.md#format). 

```suggestion
        on_output = function(params, done)
```

johnallen3d replied:
That did it, thanks so much! I poured over the docs but somehow missed that section. 🤦‍♂️

jose-elias-alvarez commented on lua/null-ls/builtins/diagnostics/cue_fmt.lua:
```diff
@@ -0,0 +1,46 @@
+                end
+            end
+
+            return diagnostics
```
```suggestion
            done(diagnostics)
```

jose-elias-alvarez reviewed (commented):
I think the changes I suggested should take care of your issue, but let me know. You'll also need to update the tests a bit to handle this structure – you can see what I did in ffbfbae1629a24851e738903b1dda45856c7d754 as an example.

johnallen3d commented:
Thanks to your guidance @jose-elias-alvarez this is working now. I've just pulled from `upstream/main`, rebased  and squashed my commits. Hopefully this is good to go now.

jose-elias-alvarez commented on doc/BUILTINS.md:
```diff
@@ -1714,6 +1714,47 @@ local sources = { null_ls.builtins.diagnostics.credo }
+
+##### About
+
+`cue fmt` formats `.cue` language files and provides formatting errors when
```
I'm sure this is an accurate description, but while reading it I had to remind myself whether this was a diagnostics or formatting source. Do you think we could change it a bit to focus on the errors?

johnallen3d replied:
Yeah, I struggled with that as well. Is the updated statement more clear?

jose-elias-alvarez replied:
Yep, much clearer IMO.

jose-elias-alvarez commented on doc/BUILTINS.md:
```diff
@@ -1714,6 +1714,47 @@ local sources = { null_ls.builtins.diagnostics.credo }
+- `command = "cue"`
+- `args = { "fmt", "$FILENAME" }`
+
+##### About
```
I think this section got mixed up with another source's documentation when you rebased.

johnallen3d replied:
Fixed.

jose-elias-alvarez commented on lua/null-ls/builtins/diagnostics/cue_fmt.lua:
```diff
@@ -0,0 +1,46 @@
+            local lines = {}
+            local diagnostics = {}
+
+            for line in params.output:gmatch("[^\r\n]+") do
```
If the goal here is to split at each newline, we could use `vim.split`.

johnallen3d replied:
Good call, much cleaner!

jose-elias-alvarez reviewed (commented):
I think there was a mix-up when rebasing the documentation but I think this is almost good to go!

johnallen3d commented on test/spec/builtins/diagnostics_spec.lua:
```diff
@@ -506,7 +506,7 @@ describe("diagnostics", function()
         end)
         it("should create a diagnostic with error severity", function()
-            local output = vim.json.decode([[ 
+            local output = vim.json.decode([[
```
My config strips trailing whitespace on save. I can revert these two 👆 changes if you prefer.

jose-elias-alvarez replied:
Not a big deal IMO.

johnallen3d commented:
@jose-elias-alvarez - thanks for the review and suggestions. I've made the requested updates however my spec is no longer passing. I'm certain the diagnostic is working because I'm using this branch locally but no matter what I do (even commenting out my new builtin) the spec reports:

```sh
Fail	||	diagnostics cue_fmt should create a diagnostic	
            test/spec/builtins/diagnostics_spec.lua:768: Expected objects to be the same.
            Passed in:
            (table: 0x01064f0088) { }
            Expected:
            (table: 0x01064f0790) {
             *[col] = 2
              [end_col] = 3
              [message] = 'expected label or ':', found 'INT' 42:'
              [row] = 3
              [severity] = 2 }
            
            stack traceback:
            	test/spec/builtins/diagnostics_spec.lua:768: in function <test/spec/builtins/diagnostics_spec.lua:765>
```

I've double-checked the specs for `credo` (the other diagnostic using `raw`) and as far as I can tell my setup matches. 🤔

jose-elias-alvarez commented:
I went ahead and fixed the test. I think there was an issue splitting the lines as well as a couple of type mismatches, but everything looks good now. Thanks for the contribution!

johnallen3d commented:
Thanks for the excellent tool and all your help @jose-elias-alvarez! Apologies for the newbie question/issues.

jose-elias-alvarez commented:
No worries, thank you for your patience!
